tests/pems/crl/generate.sh

```diff
@@ -0,0 +1,203 @@
+}
+
+init_ca_dir() {
+  ca_dir="$1"
+  mkdir -p "${ca_dir}"
+  touch "${ca_dir}/certindex"
+  echo 01 > "${ca_dir}/certserial"
+  echo 01 > "${ca_dir}/crlnumber"
+}
+
+generate_key_cert() {
+  path="$1"
+  name=$(basename "${path}")
+  openssl genrsa -out "${path}.key" 4096 || exit
+  openssl req -new \
+    -key "${path}.key" \
+    -out "${path}.csr" \
+    -subj "$(get_subj ${name})" \
+    || exit
+}
+
+generate_key_cert_self_signed() {
+  path="$1"
+  name=$(basename "${path}")
+  openssl genrsa -out "${path}.key" 4096
+  openssl req -new -x509 -days 36500 \
+    -key "${path}.key" \
+    -out "${path}.crt" \
+    -subj "$(get_subj ${name})" \
+    || exit
+}
+
+OPENSSL_CONF_FILE="$(pwd)/openssl.conf"
+
+sign_cert() {
+  path="$1"
+  ca_name="$2" openssl ca -batch -config "${OPENSSL_CONF_FILE}" -notext \
+    -in "${path}.csr" \
+    -out "${path}.crt" \
+    || exit
+}
+
+generate_crl() {
+  ca="$1"
+  ca_name="${ca}" openssl ca -config "${OPENSSL_CONF_FILE}" -gencrl \
+    -keyfile "${ca}.key" \
+    -cert "${ca}.crt" \
+    -out "${ca}.crl.pem" \
+    || exit
+  openssl crl -inform PEM \
+    -in "${ca}.crl.pem" \
+    -outform DER -out "${ca}.crl" \
+    || exit
+  rm "${ca}.crl.pem"
+}
+
+revoke_certificate() {
+  cert_path="$1"
+  ca="$2"
+  ca_name="${ca}" openssl ca -config "${OPENSSL_CONF_FILE}" \
+    -revoke "${cert_path}" \
+    -keyfile "${ca}.key" \
+    -cert "${ca}.crt" \
+    || exit
+  generate_crl "${ca}"
+}
+
+#
+# Generate certificate chains of valid and revoked certificates with 1 intermediate CA.
+#
+# root -> valid intermediate   -> valid leaf
+#                              -> revoked leaf
+#      -> revoked intermediate -> valid leaf
+#                              -> revoked leaf
+#
+
+# Generate root certificate
+init_ca_dir ./root
+generate_key_cert_self_signed ./root/root
+
+# Generate intermediate certificates
+init_ca_dir ./root/intermediate
+generate_key_cert ./root/intermediate/intermediate
+
+init_ca_dir ./root/intermediate_revoked
+generate_key_cert ./root/intermediate_revoked/intermediate_revoked
+
+# Generate leaf certificates
+mkdir ./root/intermediate/leaf
+generate_key_cert ./root/intermediate/leaf/leaf
+generate_key_cert ./root/intermediate/leaf/leaf_revoked
+
+mkdir ./root/intermediate_revoked/leaf
+generate_key_cert ./root/intermediate_revoked/leaf/leaf
+generate_key_cert ./root/intermediate_revoked/leaf/leaf_revoked
+
+# Sign intermediates with root
+pushd ./root 1> /dev/null || exit
+sign_cert ./intermediate/intermediate root
+sign_cert ./intermediate_revoked/intermediate_revoked root
+popd 1> /dev/null || exit
+
+# Sign leaves with intermediates
+pushd ./root/intermediate 1> /dev/null || exit
+sign_cert ./leaf/leaf intermediate
+sign_cert ./leaf/leaf_revoked intermediate
+popd 1> /dev/null || exit
+
+pushd ./root/intermediate_revoked 1> /dev/null || exit
+sign_cert ./leaf/leaf intermediate_revoked
+sign_cert ./leaf/leaf_revoked intermediate_revoked
+popd 1> /dev/null || exit
+
```

Comment:
Wondering if it would be useful to have a validation step where we check that the root cert can validate the intermediate cert, which can validate the leaf cert. Something like here: https://jamielinux.com/docs/openssl-certificate-authority/create-the-intermediate-pair.html#verify-the-intermediate-certificate. Just to check that the certificates you're creating are working as described.

Reply:
Thanks for the link! Following this guide made things a lot easier.

I added two new validation checks. The first is a check in `sign_cert()` that checks to make sure openssl can verify the new certificate with all of the previous ones in the chain: https://github.com/aws/s2n-tls/blob/c1615a6fcf518ac9d84564d9647d97029f3106b2/tests/pems/crl/generate.sh#L141-L142

The second check makes sure openssl can read the generated CRLs and can properly determine which certificates are rejected and which aren't: https://github.com/aws/s2n-tls/blob/c1615a6fcf518ac9d84564d9647d97029f3106b2/tests/pems/crl/generate.sh#L225-L233

Reply:
Yeah certificates are notoriously hard to generate correctly 🤣